File: Models/MessageKitModels.swift
import Foundation
import MessageKit
import CoreLocation
import FirebaseAuth

// MARK: - MessageKit Adapter Models

/// Modèle utilisateur pour MessageKit
struct MessageSender: SenderType {
    var senderId: String
    var displayName: String
    
    init(userId: String, name: String) {
        self.senderId = userId
        self.displayName = name
    }
}

/// Message adapté pour MessageKit
struct DailyQuestionMessage: MessageType {
    var sender: SenderType
    var messageId: String
    var sentDate: Date
    var kind: MessageKind
    
    init(response: QuestionResponse) {
        self.sender = MessageSender(userId: response.userId, name: response.userName)
        self.messageId = response.id
        self.sentDate = response.respondedAt
        self.kind = .text(response.text)
    }
    
    // Initializer pour créer un message temporaire
    init(tempId: String, text: String, sender: MessageSender, date: Date = Date()) {
        self.sender = sender
        self.messageId = tempId
        self.sentDate = date
        self.kind = .text(text)
    }
}

/// Service de conversion QuestionResponse ↔ MessageKit
class MessageKitAdapter {
    
    /// Convertit une QuestionResponse en DailyQuestionMessage
    static func convert(_ response: QuestionResponse) -> DailyQuestionMessage {
        return DailyQuestionMessage(response: response)
    }
    
    /// Convertit un array de QuestionResponse en messages MessageKit
    static func convert(_ responses: [QuestionResponse]) -> [DailyQuestionMessage] {
        return responses
            .sorted { $0.respondedAt < $1.respondedAt }
            .map { convert($0) }
    }
    
    /// Crée un sender pour l'utilisateur actuel
    static func currentUserSender(appState: AppState) -> MessageSender? {
        guard let currentUserId = Auth.auth().currentUser?.uid,
              let currentUser = appState.currentUser else {
            return nil
        }
        return MessageSender(userId: currentUserId, name: currentUser.name)
    }
    
    /// Crée un sender pour le partenaire
    static func partnerSender(from responses: [QuestionResponse], currentUserId: String) -> MessageSender? {
        guard let partnerResponse = responses.first(where: { $0.userId != currentUserId }) else {
            return nil
        }
        return MessageSender(userId: partnerResponse.userId, name: partnerResponse.userName)
    }
} 

File: Models/MessageKitModels.kt
package com.love2love.chat

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import java.util.Date

// =============================
// Android strings.xml resolver
// =============================
/**
 * Utilitaire pour résoudre une ressource de chaîne par son nom
 * (permet d'utiliser des clés dynamiques style "daily_challenge_1").
 * Exemple d'usage : StringsResolver.byName(context, "daily_challenge_1")
 */
object StringsResolver {
    fun byName(context: Context, key: String): String {
        val resId = context.resources.getIdentifier(key, "string", context.packageName)
        return if (resId != 0) context.getString(resId) else key // fallback sûr
    }
}

// =============================
// Modèles
// =============================

data class MessageSender(
    val senderId: String,
    val displayName: String
)

sealed class MessageKind {
    data class Text(val text: String) : MessageKind()
    // Évolutif : Image, Audio, etc.
}

data class DailyQuestionMessage(
    val sender: MessageSender,
    val messageId: String,
    val sentDate: Date,
    val kind: MessageKind
) {
    companion object {
        fun temporary(
            tempId: String,
            text: String,
            sender: MessageSender,
            date: Date = Date()
        ): DailyQuestionMessage = DailyQuestionMessage(
            sender = sender,
            messageId = tempId,
            sentDate = date,
            kind = MessageKind.Text(text)
        )
    }
}

// =============================
// Modèle minimal QuestionResponse
// (supprimez-le si vous l'avez déjà ailleurs)
// =============================

data class QuestionResponse(
    val id: String,
    val userId: String,
    val userName: String,
    val text: String,
    val respondedAt: Date,
    val status: String? = null,
    val isReadByPartner: Boolean? = null
)

// =============================
// Adapter/convertisseur
// =============================

object MessageAdapter {

    /** Convertit une QuestionResponse en DailyQuestionMessage */
    fun convert(response: QuestionResponse): DailyQuestionMessage =
        DailyQuestionMessage(
            sender = MessageSender(userId = response.userId, displayName = response.userName),
            messageId = response.id,
            sentDate = response.respondedAt,
            kind = MessageKind.Text(response.text)
        )

    /** Convertit une liste de QuestionResponse triée par date ascendante */
    fun convert(responses: List<QuestionResponse>): List<DailyQuestionMessage> =
        responses
            .sortedBy { it.respondedAt }
            .map { convert(it) }

    /** Crée un sender pour l'utilisateur courant (FirebaseAuth) */
    fun currentUserSender(currentUserDisplayName: String? = null): MessageSender? {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return null
        val name = currentUserDisplayName ?: "User"
        return MessageSender(senderId = uid, displayName = name)
    }

    /**
     * Crée un sender pour le partenaire en trouvant la première réponse dont l'userId
     * est différent de currentUserId.
     */
    fun partnerSender(responses: List<QuestionResponse>, currentUserId: String): MessageSender? {
        val partner = responses.firstOrNull { it.userId != currentUserId } ?: return null
        return MessageSender(senderId = partner.userId, displayName = partner.userName)
    }
}
